Feature 1 Move Player Use Case

iamyanbo opened the pull request.

alirezazim03 commented on src/main/java/GameEntities/Tiles/JailTile.java:
```diff
@@ -26,6 +26,6 @@ public JailTile() {
     @Override
     public TileActionResultModel action(Player player) {
-        return new TileActionResultModel(player, -1);
+        return new TileActionResultModel("You are just visiting jail", -1);
```
this flavor text should only be used when the player has landed on jail without actually being in jail

alirezazim03 commented on src/test/java/TurnUseCases/TestMovePlayerUseCase.java:
```diff
@@ -0,0 +1,61 @@
+        int[] playerRollAmount = {1, 2};
+        Player player = new Player("player", "icon", 500);
+        MovePlayerPresenter movePlayerPresenter = new MovePlayerPresenter();
+        Board board = new Board(); // Constructor will need inputs
```
make sure you update this when the Board class has been implemented!

alirezazim03 commented on src/test/java/TurnUseCases/TestMovePlayerUseCase.java:
```diff
@@ -0,0 +1,61 @@
+        int[] playerRollAmount = {1, 1};
+        Player player = new Player("player", "icon", 500);
+        MovePlayerPresenter movePlayerPresenter = new MovePlayerPresenter();
+        Board board = new Board(); // Constructor will need inputs
```
make sure you update this when the Board class has been implemented

alirezazim03 commented on src/test/java/TurnUseCases/TestMovePlayerUseCase.java:
```diff
@@ -0,0 +1,61 @@
+package TurnUseCases;
```
it might be a smart idea to add a test for when the player just lands on the jail tile not the go to jail tile

JLenander reviewed (approved):
You can rename the `startAction` methods in the input/output boundary to more explicit names if you want as it is solely you defining them. Otherwise looks good.

TahirLeonenko reviewed (changes requested):
1. In MovePlayerUseCase, for the the if branch were you want the player to move forwards, the logic is if( steps < 0), where steps = absolutePosition - player.getPosition();
If we want to move forwards, then the absolute position would be larger than player position, meaning you should be checking if steps is bigger than 0 instead of smaller.

2. This logic also ignores times when the card moves forwards through where the board resets, in these cases you're still moving forwards but the absolute position is smaller than the player position which will make your code start moving the player backwards instead.

3. You use player.updatePosition(1); but updatePosition takes in board as well, so you will need to pass in board as well (you can do this when Ali updates player to include an instance of board)

4. I was looking over which cards move backwards, and the only move 3 steps back card is in chance. None of the chance tiles are placed that you will pass backwards through go, so you can change the move backwards else branch to get rid of the nullPass.

5. You use some methods that are in the EndTurnOutputBoundary interface(showBuyableProperty, showRoll etc.), but there is no class that implements the interface and it's methods.

6. The logic that checks whether player got moved to jail (due to the card) uses player.getPosition instead of result.getPlayerPosition (the cards model returns -1 in the playerPosition attribute but the actual player position on the board is the jailTile, not -1)

7. In the else branch for normal move player cards, you use result.getTileName(), result.getTileDescription(), when you should be using result.getCardName, result.getFlavorText (the card description uses the TileResultActionModel's attribute there instead of creating a redundant new one) and result.isChance (you use result.getIsChance)

8.  You use the wrong getters in the else branch for cards that don't move the player

9. When a player rolls a double they don't get a separate turn (GameState only handles things between turns so I think the player rolling again should be handled in startAction?)

10. In the get out of jail use case you do player.enterJail if the roll that got them out of jail made them land on the GoToJail Tile (this is not possible so you can remove this if branch), but the GoToJail Tile already uses this method to send the player to jail.

11. 12 spaces from the jail til is a chance tile, you need to incorporate that into this start action (maybe you should turn the code you made in the first startAction into a separate method that these startActions call? For code clarity)

12. There's no logic for if the player has reached the maximum amount of turns in jail

iamyanbo commented:
Fixed issues:
1. Update inequality
2 and 4. Turned into 1 as There is only 1 case where the user moves backwards and that's when absPosition-playerPosition = -3
3. This change will be handled by Ali's PR's
5. The output boundaries are meant to be implemented by the presenters which we haven't made yet
6. Updated
7. Updated
8. Assuming you meant the same as 7, updated
9. The logic to roll again is not handled in the MovePlayerUseCase, it only handles rolling and moving the player, should be handled by turn controller
10. Updated
11. Updated
12. This is not handled in TryToGetOutOfJail use case as when the player reached the max turns in jail they should use the MovePlayer use case, turn controller should handle the logic for determining which use case to use.

TahirLeonenko reviewed (changes requested):
1. (moveToPosition move case) You tried fixing the cases where you pass through go (absolute position smaller than player position) by adding the size of the board (line 60), but now when player has a bigger position than absolute position you're making the player go around the board an extra time which will show up on the output as it get's updated  at each updatePosition(1) and it also makes player get go money when they shouldn't

2. In the get out of jail use case, the else branch (line 64) that's supposed to handle a normal move player card doesn't call the moveToPosition method in MovePlayerUseCase, which is what moves the player for normal AdvanceCards

Youssefb1234 commented on src/main/java/TurnUseCases/MovePlayerUseCase/MovePlayerUseCase.java:
```diff
@@ -0,0 +1,150 @@
+    }
+
+    @Override
+    public void moveToPosition(Player player, int absolutePosition) {
```
Although this method appears functionally adequate, its implementation can be much cleaner as it suffers from some code smells (namely long method, and too many data clumps). After doing some unit tests for the methods, I suggest code refactoring by fragmenting the code and eliminating any redundancies. If there aren't any, then maybe rearrange and focus on concision.

alirezazim03 reviewed (approved):
good
